tests/e2e/messages-content.js

```diff
@@ -0,0 +1,209 @@
+        ])
+        .then(results => {
+          // My kingdom for results.reduce(&&);
+          return results[0] && results[1] && results[2];
+        });
+    }, 10000);
+
+    return findSelect2Entry(entrySelector, entryText).first();
+  };
+
+  const enterCheckAndSelect = (
+    searchText,
+    totalExpectedResults,
+    entrySelector,
+    entryText,
+    existingEntryCount
+  ) => {
+    existingEntryCount = existingEntryCount || 0;
+    expect(element.all(by.css('li.select2-selection__choice')).count()).toBe(
+      existingEntryCount
+    );
+
+    const entry = searchSelect2(
+      searchText,
+      totalExpectedResults,
+      entrySelector,
+      entryText
+    );
+    entry.click();
+
+    browser.wait(() => {
+      return element
+        .all(by.css('li.select2-selection__choice'))
+        .count()
+        .then(utils.countOf(existingEntryCount + 1));
+    }, 2000);
+  };
+
+  const sendMessage = () => {
+    element(by.css('#send-message a.btn.submit:not(.ng-hide)')).click();
+
+    browser.wait(() => {
+      return element(by.css('#send-message'))
+        .isDisplayed()
+        .then(isDisplayed => {
+          return !isDisplayed;
+        })
+        .catch(() => {
+          // It's been detached, so it's gone
+          return true;
+        });
+    }, 2000);
+
+    utils.resetBrowser();
+  };
+
+  const clickLhsEntry = (entryId, entryName) => {
+    entryName = entryName || entryId;
+    const liIdentifier = messageInList(entryId);
+
+    helper.waitUntilReady(element(by.css(liIdentifier)));
+
+    expect(element.all(by.css(liIdentifier)).count()).toBe(1);
+    element(by.css(liIdentifier + ' a')).click();
+
+    browser.wait(() => {
+      const el = element(by.css('#message-header .name'));
+      if (helper.waitUntilReady(el)) {
+        return helper.getTextFromElement(el).then(text => {
+          return text === entryName;
+        });
+      }
+    }, 12000);
+  };
+
+  const contactNameSelector = ' .sender .name';
+
+  const getElementText = (css, attempt) => {
+    attempt = attempt || 0;
+
+    return helper.getTextFromElement(element(by.css(css)))
+      .then((text) => {
+        return text;
+      }, (err) => {
+        if (attempt < 2) {
+          return getElementText(css, attempt+1);
+        }
+        throw err;
+      });
+  };
+
+  describe('Display message without contact', () => {
+    it('can display messages without contact', () => {
+      common.goToMessages();
+
+      expect(element(by.css(messageInList(ALICE._id))).isPresent()).toBeFalsy();
+
+      openSendMessageModal();
+      enterCheckAndSelect(ALICE.name, 2, contactNameSelector, ALICE.name);
+      element(by.css('#send-message textarea')).sendKeys(smsMsg('contact'));
+      sendMessage();
+      clickLhsEntry(ALICE._id, ALICE.name);
+
+      browser.wait(() => {
+        return utils.deleteDocs(CONTACTS.map(contact => contact._id));
+      });
+
+      browser.sleep(1000);
```

Comment:
Why do you need this sleep here?

Using sleep generally is a really bad idea, because it makes your test entirely dependent on how fast the executing computer is (maybe 1000ms isn't enough on all Travis nodes? Or on someone's crappy laptop..), as well as slowing the test down for everyone whose computer **is** fast enough.

I don't think we have a general rule about this, but I'd like to see either the sleep removed (ie detect and wait for the right page change so it always works) or a comment explaining what is happening while we're sleeping and why it needs to be a sleep